SDK image = runtime image + development utilities

endersonmaia opened the pull request:
This pull request includes several changes to the `@cartesi/sdk` package, focusing on updating Docker images, adding new workflows, and modifying the Dockerfile. The most important changes include adding new container image releases, updating base images, creating databases with migrations, and adding new GitHub workflows for building and pushing Docker images.

### Updates to Docker images and base images:

* Updated the Debian base image to `bookworm-20250317-slim` in `.changeset/honest-spoons-appear.md` and `packages/sdk/docker-bake.hcl`. [[1]](diffhunk://#diff-15eceb9afa3481da404dacc17863b1efb4c5ef2ac09556324df8af8a2686285bR1-R5) [[2]](diffhunk://#diff-f31033306f300af23520856c404681eb6cbe45d9f28ce0f28abb24fcae811e5bL8-R35)
* Added new container image releases for `cartesi/rollups-database` and updated the Dockerfile to use new base images for `cartesi`, `espresso-dev-node`, and `postgres`. [[1]](diffhunk://#diff-a8cd29f751915a2e060ce14cffed4c65e2198a196bfcd173355571ef154ffaf0R1-R5) [[2]](diffhunk://#diff-fbbb6309bd1c5bde85dc9baab5c5c1ef5af52fedb5e3fd28cecf37916d3ef1feL2-R12) [[3]](diffhunk://#diff-fbbb6309bd1c5bde85dc9baab5c5c1ef5af52fedb5e3fd28cecf37916d3ef1feR123-R254) [[4]](diffhunk://#diff-fbbb6309bd1c5bde85dc9baab5c5c1ef5af52fedb5e3fd28cecf37916d3ef1feR266-R270) [[5]](diffhunk://#diff-fbbb6309bd1c5bde85dc9baab5c5c1ef5af52fedb5e3fd28cecf37916d3ef1feL191-R320)

### Database migrations:

* Created databases and ran migrations for a preinitialized PostgreSQL database in `.changeset/hot-timers-repair.md`.
* Modified the Dockerfile to include scripts for creating and migrating databases for `rollupsdb`, `graphql`, and `espresso`.

### New GitHub workflows:

* Added `rollups-database.yml` and `runtime.yml` workflows to build and push Docker images for `rollups-database` and `sdk-node`. [[1]](diffhunk://#diff-f94c5962e154f5f3cda6559fe424eaa26ec86766ca1a928017cd438822cb98b1R1-R91) [[2]](diffhunk://#diff-8625da419f84b5aea1b452b7ef296bdc86ff61521091437a2dd985787b9c3cb6R1-R91)

### Dockerfile modifications:

* Refactored the Dockerfile to use new ARGs for base images and added steps to install necessary packages and set up the environment. [[1]](diffhunk://#diff-fbbb6309bd1c5bde85dc9baab5c5c1ef5af52fedb5e3fd28cecf37916d3ef1feL2-R12) [[2]](diffhunk://#diff-fbbb6309bd1c5bde85dc9baab5c5c1ef5af52fedb5e3fd28cecf37916d3ef1feR123-R254) [[3]](diffhunk://#diff-fbbb6309bd1c5bde85dc9baab5c5c1ef5af52fedb5e3fd28cecf37916d3ef1feR266-R270) [[4]](diffhunk://#diff-fbbb6309bd1c5bde85dc9baab5c5c1ef5af52fedb5e3fd28cecf37916d3ef1feL191-R320) [[5]](diffhunk://#diff-fbbb6309bd1c5bde85dc9baab5c5c1ef5af52fedb5e3fd28cecf37916d3ef1feR338-L262)

### Docker bake configuration:

* Updated `docker-bake.hcl` and `docker-bake.override.hcl` to include new targets for `runtime` and `database`, and set appropriate tags for the images. [[1]](diffhunk://#diff-f31033306f300af23520856c404681eb6cbe45d9f28ce0f28abb24fcae811e5bL8-R35) [[2]](diffhunk://#diff-f1ace0abad9a2c988721bc088f4092fe2ec8c67b92ff5edb6f13758da3dc9bd8R4-R11)

endersonmaia committed:
chore(sdk): bump Debian base image to bookworm-20250317-slim

endersonmaia committed:
feat(sdk): add cartesi/rollups-node container release

A container image will be released only with the binaries needed to run
a rollups-node with optional auxiliary services like rollups-graphql and
espresso-reader.

tuler commented on packages/sdk/Dockerfile:
```diff
@@ -137,40 +136,115 @@ ENV POSTGRES_PASSWORD=password
+cartesi-machine --version-json
+EOF
+
+# Install linux kernel image
```
the kernel image and linux headers are only used during development

endersonmaia committed:
refactor(sdk): add CARTESI_ prefix to variables

endersonmaia committed:
refactor(sdk): convention _BASE_IMAGE prefixed ARGs with full image name

endersonmaia committed:
feat(sdk): pin container image digests

tuler commented on .github/workflows/rollups-node.yml:
```diff
@@ -0,0 +1,91 @@
+              uses: docker/metadata-action@902fa8ec7d6ecbf8d84d538b9b233a880e428804 # v5.7.0
+              with:
+                  images: |
+                      docker.io/cartesi/sdk-node,enable=${{ github.event_name != 'pull_request' }}
```
wrong name

tuler commented on packages/sdk/docker-bake.hcl:
```diff
@@ -5,22 +5,27 @@ target "default" {
     CARTESI_ROLLUPS_GRAPHQL_VERSION   = "2.3.8"
+    CARTESI_ROLLUPS_NODE_BASE_IMAGE   = "docker.io/library/debian:bookworm-20250317-slim@sha256:1209d8fd77def86ceb6663deef7956481cc6c14a25e1e64daec12c0ceffcc19d"
     CARTESI_ROLLUPS_NODE_VERSION      = "2.0.0-alpha.1"
+    CARTESI_SDK_BASE_IMAGE            = "docker.io/library/node:22-bookworm-slim@sha256:bac8ff0b5302b06924a5e288fb4ceecef9c8bb0bb92515985d2efdc3a2447052"
```
Isn't it better to use the runtime image as base?

endersonmaia replied:
Comparing the size, the difference is negligible, I sent a commit that uses a single base image.

So we have what you suggested, base -> runtime -> sdk.

tuler commented on packages/sdk/docker-bake.override.hcl:
```diff
@@ -1,3 +1,7 @@
 }
+
+target "rollups-node" {
+  tags = ["cartesi/sdk-node:devel"]
```
let's discuss a bit more on this name

tuler commented on packages/sdk/Dockerfile:
```diff
@@ -132,20 +135,82 @@ EOF
+
+COPY --from=go-migrate /usr/local/bin/migrate /usr/local/bin/
+COPY --from=graphql /usr/local/bin/cartesi-rollups-graphql /usr/local/bin/
+COPY --from=graphql-migration /usr/share/cartesi/rollups-graphql/migrations /usr/share/cartesi/rollups-graphql/migrations
```
I think migrations belongs to dev.
On production we'll probably prepare the database once separately.

tuler replied:
same related to `go-migrate`

tuler reviewed (commented):
I think the order should be something like:

```
FROM debian as runtime
# add cartesi-machine
# add cartesi-rollups-node
# add graphql
# add espresso-reader

FROM runtime as sdk
# add anvil
# add devnet
# add crane
# add genext2fsx
# add migrations
```

endersonmaia commented:
> I think the order should be something like:
> 
> ```
> FROM debian as runtime
> # add cartesi-machine
> # add cartesi-rollups-node
> # add graphql
> # add espresso-reader
> 
> FROM runtime as sdk
> # add anvil
> # add devnet
> # add crane
> # add genext2fsx
> # add migrations
> ```

Yeah, I decided to use `node:slim` as the base for the sdk to reduce the size a little bit.

The order you suggested makes the code more clean at the expense of some MB in the final sdk image.

endersonmaia committed:
refactor(sdk): single base image

endersonmaia committed:
feat(sdk): add cartesi/rollups-database container release

This container image will be based on postgres and contains a
pre-initialized database with rollups-node, rollups-graphql and espresso-reader
database migrations ready to be used.

endersonmaia commented:
> > I think the order should be something like:
> > ```
> > FROM debian as runtime
> > # add cartesi-machine
> > # add cartesi-rollups-node
> > # add graphql
> > # add espresso-reader
> > 
> > FROM runtime as sdk
> > # add anvil
> > # add devnet
> > # add crane
> > # add genext2fsx
> > # add migrations
> > ```
> 
> Yeah, I decided to use `node:slim` as the base for the sdk to reduce the size a little bit.
> 
> The order you suggested makes the code more clean at the expense of some MB in the final sdk image.

Just accepted your suggestion after evaluating the size difference was negligible.

endersonmaia commented:
We still need to define the naming, and we make the final fixups to stick with whatever naming is decided.

tuler commented on .github/workflows/rollups-database.yaml:
```diff
@@ -0,0 +1,91 @@
+name: rollups-database
```
use `database` and `database.yaml` to follow name convention.

tuler commented on packages/sdk/Dockerfile:
```diff
@@ -1,25 +1,15 @@
 # base image
-FROM ${BASE_IMAGE} AS base
+FROM ${CARTESI_BASE_IMAGE} AS base
 SHELL ["/bin/bash", "-euo", "pipefail", "-c"]
```
do we need this SHELL command?

endersonmaia replied:
This is just to make the `RUN <<EOF` run with `set -euo pipefail` by default, without the need to repeat it.

tuler commented on packages/sdk/Dockerfile:
```diff
@@ -130,28 +120,138 @@ curl -fsSL "https://github.com/cartesi/rollups-espresso-reader/archive/refs/tags
+    cartesi
+EOF
+
+RUN mkdir -p /tmp/.cartesi && chmod 1777 /tmp/.cartesi
```
I think we don't need this anymore

tuler commented on packages/sdk/Dockerfile:
```diff
@@ -130,28 +120,138 @@ curl -fsSL "https://github.com/cartesi/rollups-espresso-reader/archive/refs/tags
+    -o /tmp/cartesi-machine.deb
+dpkg -i /tmp/cartesi-machine.deb
+rm /tmp/cartesi-machine.deb
+cartesi-machine --version-json
```
do we need this call?

endersonmaia replied:
This is a smoke-test during build.

tuler commented on packages/sdk/Dockerfile:
```diff
@@ -130,28 +120,138 @@ curl -fsSL "https://github.com/cartesi/rollups-espresso-reader/archive/refs/tags
+mkdir -p /var/lib/cartesi-rollups-node/snapshots
+chmod 755 /var/lib/cartesi-rollups-node/snapshots
+chown cartesi:cartesi /var/lib/cartesi-rollups-node/snapshots
+cartesi-rollups-node --version
```
do we need this call?

endersonmaia replied:
same as last comment

tuler commented on packages/sdk/Dockerfile:
```diff
@@ -130,28 +120,138 @@ curl -fsSL "https://github.com/cartesi/rollups-espresso-reader/archive/refs/tags
+EOF
+
+COPY --from=runtime /usr/bin/cartesi-rollups-cli /usr/bin/
+COPY --from=runtime /usr/lib/libcartesi* /usr/lib/
```
is this safe? why is it needed?

endersonmaia replied:
We need this to run `cartesi-rollups-cli` without the need to reinstall its package.

tuler commented on packages/sdk/Dockerfile:
```diff
@@ -130,28 +120,138 @@ curl -fsSL "https://github.com/cartesi/rollups-espresso-reader/archive/refs/tags
 USER root
 ARG DEBIAN_FRONTEND=noninteractive
 RUN <<EOF
+apt-get update
```
some of the packages below were already installed by runtime base

tuler commented:
Here is a commit organization that may be better to review and merge:

A commit with only the Dockerfile. No other change is necessary to the workflows, as the final default target is the sdk as it is today. That is already mergable by itself.

Then we handle the case for `database` and `runtime` on top of that.
I have an idea for the workflow that might simplify it.

endersonmaia committed:
fixup! feat(sdk): add cartesi/rollups-database container release

endersonmaia committed:
fixup! feat(sdk): add cartesi/rollups-node container release

endersonmaia commented:
> A commit with only the Dockerfile. No other change is necessary to the workflows, as the final default target is the sdk as it is today. That is already mergable by itself.

There's a PR attempt for that at #207

endersonmaia commented:
> Then we handle the case for `database` and `runtime` on top of that.

Another PR for that: https://github.com/cartesi/cli/pull/209

tuler commented:
Superseded by https://github.com/cartesi/cli/pull/209
